src/types/string.rs

```diff
@@ -1,14 +1,129 @@
 // Copyright (c) 2017-present PyO3 Project and Contributors
 
+#[cfg(not(Py_LIMITED_API))]
+use crate::exceptions::PyUnicodeDecodeError;
 use crate::types::PyBytes;
 use crate::{
     ffi, AsPyPointer, FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python,
     ToPyObject,
 };
 use std::borrow::Cow;
+#[cfg(not(Py_LIMITED_API))]
+use std::ffi::CStr;
 use std::os::raw::c_char;
 use std::str;
 
+/// Represents raw data backing a Python `str`.
+///
+/// Python internally stores strings in various representations. This enumeration
+/// represents those variations.
+#[cfg(not(Py_LIMITED_API))]
+#[cfg_attr(docsrs, doc(cfg(not(Py_LIMITED_API))))]
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PyStringData<'a> {
+    /// UCS1 representation.
+    Ucs1(&'a [u8]),
+
+    /// UCS2 representation.
+    Ucs2(&'a [u16]),
+
+    /// UCS4 representation.
+    Ucs4(&'a [u32]),
+}
+
+#[cfg(not(Py_LIMITED_API))]
+impl<'a> PyStringData<'a> {
+    /// Obtain the raw bytes backing this instance as a [u8] slice.
+    pub fn as_bytes(&self) -> &[u8] {
+        match self {
+            Self::Ucs1(s) => s,
+            Self::Ucs2(s) => unsafe {
+                std::slice::from_raw_parts(
+                    s.as_ptr() as *const u8,
+                    s.len() * self.character_width_bytes(),
+                )
+            },
+            Self::Ucs4(s) => unsafe {
+                std::slice::from_raw_parts(
+                    s.as_ptr() as *const u8,
+                    s.len() * self.character_width_bytes(),
+                )
+            },
+        }
+    }
+
+    /// Size in bytes of each *character* in the underlying buffer.
```

Comment:
Is "character" the correct terminology here? IMO it should be "codepoint".

Reply:
It is debatable.

It is unclear to me if Python's internal storage format actually guarantees that a slice N byte value is the same as a _code point_. UTF-8 and UTF-16 are variable width encodings after all. I _think_ Python will do things like store surrogate pairs without expanding the item width of the stored value.

The documentation at https://docs.python.org/3/c-api/unicode.html uses _character_ to refer to the values in these buffers, which is the main reason I used _character_ in the API.

But, I don't likely blindly copying things, especially if they are wrong.

What I know for certain is that the values stored in the memory slice are stored with a fixed width. So maybe we should go with a name that says what they actually are: `item width in bytes`.

🤷‍♂️